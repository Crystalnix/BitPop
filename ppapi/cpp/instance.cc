// Copyright (c) 2011 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "ppapi/cpp/instance.h"

#include "ppapi/c/dev/ppp_printing_dev.h"
#include "ppapi/c/ppb_instance.h"
#include "ppapi/c/ppb_messaging.h"
#include "ppapi/cpp/common.h"
#include "ppapi/cpp/dev/surface_3d_dev.h"
#include "ppapi/cpp/graphics_2d.h"
#include "ppapi/cpp/image_data.h"
#include "ppapi/cpp/logging.h"
#include "ppapi/cpp/module.h"
#include "ppapi/cpp/module_impl.h"
#include "ppapi/cpp/point.h"
#include "ppapi/cpp/resource.h"
#include "ppapi/cpp/var.h"

namespace pp {

namespace {

template <> const char* interface_name<PPB_Instance>() {
  return PPB_INSTANCE_INTERFACE;
}

template <> const char* interface_name<PPB_Messaging>() {
  return PPB_MESSAGING_INTERFACE;
}

}  // namespace

Instance::Instance(PP_Instance instance) : pp_instance_(instance) {
}

Instance::~Instance() {
  // Ensure that all per-instance objects have been removed. Generally, these
  // objects should have their lifetime scoped to the instance, such as being
  // instance members or even implemented by your instance sub-class directly.
  //
  // If they're not unregistered at this point, they will usually have a
  // dangling reference to the instance, which can cause a crash later.
  PP_DCHECK(interface_name_to_objects_.empty());
}

bool Instance::Init(uint32_t /*argc*/, const char* /*argn*/[],
                    const char* /*argv*/[]) {
  return true;
}

void Instance::DidChangeView(const pp::Rect& /*position*/,
                             const pp::Rect& /*clip*/) {
}

void Instance::DidChangeFocus(bool /*has_focus*/) {
}


bool Instance::HandleDocumentLoad(const URLLoader& /*url_loader*/) {
  return false;
}

bool Instance::HandleInputEvent(const PP_InputEvent& /*event*/) {
  return false;
}

void Instance::HandleMessage(const Var& /*message*/) {
  return;
}

Var Instance::GetSelectedText(bool /* html */) {
  return Var();
}

#ifndef PPAPI_INSTANCE_REMOVE_SCRIPTING
Var Instance::GetWindowObject() {
  if (!has_interface<PPB_Instance>())
    return Var();
  return Var(Var::PassRef(),
             get_interface<PPB_Instance>()->GetWindowObject(pp_instance()));
}

Var Instance::GetOwnerElementObject() {
  if (!has_interface<PPB_Instance>())
    return Var();
  return Var(Var::PassRef(),
             get_interface<PPB_Instance>()->GetOwnerElementObject(
                 pp_instance()));
}

Var Instance::ExecuteScript(const Var& script, Var* exception) {
  if (!has_interface<PPB_Instance>())
    return Var();
  return Var(Var::PassRef(),
             get_interface<PPB_Instance>()->ExecuteScript(
                 pp_instance(),
                 script.pp_var(),
                 Var::OutException(exception).get()));
}

Var Instance::GetInstanceObject() {
  return Var();
}
#endif

bool Instance::BindGraphics(const Graphics2D& graphics) {
  if (!has_interface<PPB_Instance>())
    return false;
  return PPBoolToBool(get_interface<PPB_Instance>()->BindGraphics(
      pp_instance(), graphics.pp_resource()));
}

bool Instance::BindGraphics(const Surface3D_Dev& graphics) {
  if (!has_interface<PPB_Instance>())
    return false;
  return PPBoolToBool(get_interface<PPB_Instance>()->BindGraphics(
      pp_instance(), graphics.pp_resource()));
}

bool Instance::IsFullFrame() {
  if (!has_interface<PPB_Instance>())
    return false;
  return PPBoolToBool(get_interface<PPB_Instance>()->IsFullFrame(
      pp_instance()));
}

void Instance::PostMessage(const Var& message) {
  if (!has_interface<PPB_Messaging>())
    return;
  get_interface<PPB_Messaging>()->PostMessage(pp_instance(),
                                              message.pp_var());
}

void Instance::AddPerInstanceObject(const std::string& interface_name,
                                    void* object) {
  // Ensure we're not trying to register more than one object per interface
  // type. Otherwise, we'll get confused in GetPerInstanceObject.
  PP_DCHECK(interface_name_to_objects_.find(interface_name) ==
            interface_name_to_objects_.end());
  interface_name_to_objects_[interface_name] = object;
}

void Instance::RemovePerInstanceObject(const std::string& interface_name,
                                       void* object) {
  InterfaceNameToObjectMap::iterator found = interface_name_to_objects_.find(
      interface_name);
  if (found == interface_name_to_objects_.end()) {
    // Attempting to unregister an object that doesn't exist or was already
    // unregistered.
    PP_DCHECK(false);
    return;
  }

  // Validate that we're removing the object we thing we are.
  PP_DCHECK(found->second == object);
  (void)object;  // Prevent warning in release mode.

  interface_name_to_objects_.erase(found);
}

// static
void* Instance::GetPerInstanceObject(PP_Instance instance,
                                     const std::string& interface_name) {
  Instance* that = Module::Get()->InstanceForPPInstance(instance);
  if (!that)
    return NULL;
  InterfaceNameToObjectMap::iterator found =
      that->interface_name_to_objects_.find(interface_name);
  if (found == that->interface_name_to_objects_.end())
    return NULL;
  return found->second;
}

}  // namespace pp
