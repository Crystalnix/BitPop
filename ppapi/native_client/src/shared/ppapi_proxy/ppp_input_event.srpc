# Copyright (c) 2011 The Native Client Authors. All rights reserved.
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.

# RPC methods used to implement the PPP_InputEvent interface.
# See ppapi/c/ppp_input_event.h for interface details.
{
 'name': 'PppInputEventRpc',
 'rpcs': [
          {'name': 'PPP_InputEvent_HandleInputEvent',
           'inputs': [['instance', 'PP_Instance'],
                      ['resource', 'PP_Resource'],
                      ['event_data', 'char[]'],  # InputEventData
                      ['character_text', 'char[]'] # character text
                     ],
           'outputs': [['handled', 'int32_t'] # PP_Bool
                      ]
          },
         ]
}
