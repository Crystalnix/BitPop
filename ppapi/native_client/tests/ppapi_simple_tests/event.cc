// Copyright (c) 2011 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include <stdint.h>
#include <stdlib.h>

#include <sstream>
#include <string>
#include <queue>

#include "ppapi/c/pp_input_event.h"
#include "ppapi/c/ppp_input_event.h"
#include "ppapi/cpp/instance.h"
#include "ppapi/cpp/module.h"
#include "ppapi/cpp/point.h"
#include "ppapi/cpp/var.h"
#include "ppapi/cpp/input_event.h"


using std::string;
using std::queue;
using std::ostringstream;

const int kDefaultEventBufferSize = 10;

namespace {

string ModifierToString(uint32_t modifier) {
  string s;
  if (modifier & PP_INPUTEVENT_MODIFIER_SHIFTKEY) {
    s += "shift ";
  }
  if (modifier & PP_INPUTEVENT_MODIFIER_CONTROLKEY) {
    s += "ctrl ";
  }
  if (modifier & PP_INPUTEVENT_MODIFIER_ALTKEY) {
    s += "alt ";
  }
  if (modifier & PP_INPUTEVENT_MODIFIER_METAKEY) {
    s += "meta ";
  }
  if (modifier & PP_INPUTEVENT_MODIFIER_ISKEYPAD) {
    s += "keypad ";
  }
  if (modifier & PP_INPUTEVENT_MODIFIER_ISAUTOREPEAT) {
    s += "autorepeat ";
  }
  if (modifier & PP_INPUTEVENT_MODIFIER_LEFTBUTTONDOWN) {
    s += "left-button-down ";
  }
  if (modifier & PP_INPUTEVENT_MODIFIER_MIDDLEBUTTONDOWN) {
    s += "middle-button-down ";
  }
  if (modifier & PP_INPUTEVENT_MODIFIER_RIGHTBUTTONDOWN) {
    s += "right-button-down ";
  }
  if (modifier & PP_INPUTEVENT_MODIFIER_CAPSLOCKKEY) {
    s += "caps-lock ";
  }
  if (modifier & PP_INPUTEVENT_MODIFIER_NUMLOCKKEY) {
    s += "num-lock ";
  }
  return s;
}


string MouseButtonToString(PP_InputEvent_MouseButton button) {
  switch (button) {
    case PP_INPUTEVENT_MOUSEBUTTON_NONE:
      return "None";
    case PP_INPUTEVENT_MOUSEBUTTON_LEFT:
      return "Left";
    case PP_INPUTEVENT_MOUSEBUTTON_MIDDLE:
      return "Middle";
    case PP_INPUTEVENT_MOUSEBUTTON_RIGHT:
      return "Right";
    default:
      ostringstream stream;
      stream << "Unrecognized (" << static_cast<int32_t>(button) << ")";
      return stream.str();
  }
}


string KeyEvent(const pp::KeyboardInputEvent& key_event,
                const string& kind) {
  ostringstream stream;
  stream << "Key event:" << kind
         << " modifier:" << ModifierToString(key_event.GetModifiers())
         << " key_code:" << key_event.GetKeyCode()
         << " time:" << key_event.GetTimeStamp()
         << " text:" << key_event.GetCharacterText().DebugString()
         << "\n";
  return stream.str();
}


string MouseEvent(const pp::MouseInputEvent& mouse_event,
                  const string& kind) {
  ostringstream stream;
  stream << "Mouse event:" << kind
         << " modifier:" << ModifierToString(mouse_event.GetModifiers())
         << " button:" << MouseButtonToString(mouse_event.GetButton())
         << " x:" << mouse_event.GetPosition().x()
         << " y:" << mouse_event.GetPosition().y()
         << " click_count:" << mouse_event.GetClickCount()
         << " time:" << mouse_event.GetTimeStamp()
         << "\n";
  return stream.str();
}


string WheelEvent(const pp::WheelInputEvent& wheel_event) {
  ostringstream stream;
  stream << "Wheel event."
         << " modifier:" << ModifierToString(wheel_event.GetModifiers())
         << " deltax:" << wheel_event.GetDelta().x()
         << " deltay:" << wheel_event.GetDelta().y()
         << " wheel_ticks_x:" << wheel_event.GetTicks().x()
         << " wheel_ticks_y:" << wheel_event.GetTicks().y()
         << " scroll_by_page:"
         << (wheel_event.GetScrollByPage() ? "true" : "false")
         << "\n";
  return stream.str();
}


string EventToString(const pp::InputEvent& event) {
  ostringstream stream;
  switch (event.GetType()) {
    default:
    case PP_INPUTEVENT_TYPE_UNDEFINED:
     stream << "Unrecognized Event (" << static_cast<int32_t>(event.GetType())
             << ")";
      return stream.str();

    case PP_INPUTEVENT_TYPE_MOUSEDOWN:
      return MouseEvent(pp::MouseInputEvent(event), "Down");
    case PP_INPUTEVENT_TYPE_MOUSEUP:
      return MouseEvent(pp::MouseInputEvent(event), "Up");
    case PP_INPUTEVENT_TYPE_MOUSEMOVE:
      return MouseEvent(pp::MouseInputEvent(event), "Move");
    case PP_INPUTEVENT_TYPE_MOUSEENTER:
      return MouseEvent(pp::MouseInputEvent(event), "Enter");
    case PP_INPUTEVENT_TYPE_MOUSELEAVE:
      return MouseEvent(pp::MouseInputEvent(event), "Leave");

    case PP_INPUTEVENT_TYPE_WHEEL:
      return WheelEvent(pp::WheelInputEvent(event));

    case PP_INPUTEVENT_TYPE_RAWKEYDOWN:
      return KeyEvent(pp::KeyboardInputEvent(event), "RawKeyDown");
    case PP_INPUTEVENT_TYPE_KEYDOWN:
      return KeyEvent(pp::KeyboardInputEvent(event), "Down");
    case PP_INPUTEVENT_TYPE_KEYUP:
      return KeyEvent(pp::KeyboardInputEvent(event), "Up");
    case PP_INPUTEVENT_TYPE_CHAR:
      return KeyEvent(pp::KeyboardInputEvent(event), "Char");
  }
}

void StringReplace(string* input,
                   const string& find,
                   const string& replace) {
  if (find.length() == 0 || input->length() == 0) {
    return;
  }

  size_t start_pos = 0;
  while (1) {
    start_pos = input->find(find, start_pos);
    if (start_pos == string::npos) {
      break;
    }
    input->replace(start_pos, find.length(), replace);
    start_pos += replace.length();
  }
}

}  // namespace


class MyInstance : public pp::Instance {
 private:
  size_t max_buffer_size_;
  queue<string> event_buffer_;

  void ParseArgs(uint32_t argc, const char* argn[], const char* argv[]) {
     for (uint32_t i = 0; i < argc; ++i) {
       const std::string tag = argn[i];
       if (tag == "buffer_size") max_buffer_size_ = strtol(argv[i], 0, 0);
       // ignore other tags
     }
  }

  // Dump all the event via PostMessage for testing
  void FlushEventBuffer() {
      while (event_buffer_.size() > 0) {
        string s = event_buffer_.front();
        event_buffer_.pop();
        // Replace space with underscore to simplify testing
        StringReplace(&s, " ", "_");
        pp::Var message(s);
        PostMessage(message);
      }
  }

 public:
  explicit MyInstance(PP_Instance instance)
    : pp::Instance(instance), max_buffer_size_(kDefaultEventBufferSize) {
    RequestInputEvents(PP_INPUTEVENT_CLASS_MOUSE | PP_INPUTEVENT_CLASS_WHEEL);
    RequestFilteringInputEvents(PP_INPUTEVENT_CLASS_KEYBOARD);
  }

  virtual ~MyInstance() {}

  virtual bool Init(uint32_t argc, const char* argn[], const char* argv[]) {
    ParseArgs(argc, argn, argv);
    return true;
  }

  virtual bool HandleInputEvent(const pp::InputEvent& event) {
     ostringstream stream;
     stream << pp_instance() << ": " << EventToString(event);
     event_buffer_.push(stream.str());
    if (event_buffer_.size() >= max_buffer_size_) {
      FlushEventBuffer();
    }
    return true;
  }

  virtual void DidChangeFocus(bool has_focus) {
    FlushEventBuffer();
  }
};

// standard boilerplate code below
class MyModule : public pp::Module {
 public:
  virtual pp::Instance* CreateInstance(PP_Instance instance) {
    return new MyInstance(instance);
  }
};

namespace pp {
  Module* CreateModule() {
    return new MyModule();
  }
}
