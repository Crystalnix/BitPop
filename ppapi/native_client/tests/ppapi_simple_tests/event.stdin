# define a bunch of variables for readability

set_variable MODULE i(4444)
set_variable INSTANCE i(5555)
set_variable EVENT i(6666)

# NOTE: the event interface between the nexe and proxy is not strictly
#       following the ppapi model because of performance reasons.
#
# The nexe entry point is
#   PPP_InputEvent_HandleInputEvent:iiCC:i
# Where the arguments are:
#   in:
#     PP_Instance instance:
#     PP_Resource resource:
#     InputEventData event_data:
#     char[] character_text:
#   out:
#     int handled:
#
# InputEventData is described in src/shared/ppapi_proxy/input_event_data.h as:
#
# struct InputEventData {
#   InputEventData();
#   ~InputEventData();
#
#   PP_InputEvent_Type event_type;
#   uint32_t event_modifiers;
#   PP_TimeTicks event_time_stamp;
#
#   PP_InputEvent_MouseButton mouse_button;
#   int32_t mouse_click_count;
#   PP_Point mouse_position;
#   PP_Point mouse_movement;
#
#   PP_FloatPoint wheel_delta;
#   PP_FloatPoint wheel_ticks;
#   PP_Bool wheel_scroll_by_page;
#
#   uint32_t key_code;
# };

# ieee double
set_variable TIME_1 8:0x3ff0000000000000
set_variable TIME_2 8:0x3ff1000000000000
set_variable TIME_3 8:0x3ff2000000000000
set_variable TIME_4 8:0x3ff3000000000000
set_variable TIME_5 8:0x3ff4000000000000
set_variable TIME_6 8:0x3ff5000000000000
set_variable TIME_7 8:0x3ff6000000000000

# pair of ints
set_variable COORDS 4:264,4:100
set_variable MOVES  4:0,4:0

set_variable EV_MOUSE_DOWN  R(64,4:0,4:64,${TIME_1},4:0,4:1,${COORDS},${MOVES},8:0,8:0,4:0,4:0)
set_variable EV_MOUSE_UP    R(64,4:1,4:64,${TIME_2},4:0,4:0,${COORDS},${MOVES},8:0,8:0,4:0,4:0)

set_variable EV_MOUSE_MOVE  R(64,4:2,4:0,${TIME_3},4:0,4:0,${COORDS},${MOVES},8:0,8:0,4:0,4:0)

set_variable EV_MOUSE_ENTER R(64,4:3,4:0,${TIME_4},4:0,4:0,${COORDS},${MOVES},8:0,8:0,4:0,4:0)
set_variable EV_MOUSE_LEAVE R(64,4:4,4:0,${TIME_5},4:0,4:0,${COORDS},${MOVES},8:0,8:0,4:0,4:0)

set_variable EV_KEY_DOWN    R(64,4:7,4:0,${TIME_6},4:0,4:0,8:0,8:0,8:0,8:0,4:0,4:0x54)
set_variable EV_KEY_UP      R(64,4:8,4:0,${TIME_7},4:0,4:0,8:0,8:0,8:0,8:0,4:0,4:0x54)

set_variable JS_STR 4:5
set_variable JS_UND 4:0
set_variable VAR_UNDEF R(8,${JS_UND},4:0)
set_variable VAR_CHAR R(16,${JS_STR},4:1,8:0x56)


echo
echo "*** INIT MODULE"
replay 1 PPB_GetInterface:s:i  s("PPB_Core;1.0") * i(1)
replay 1 PPB_GetInterface:s:i  s("PPB_Instance;1.0") * i(1)
# the results contain a process id which is non-determinisitic
rpc PPP_InitializeModule hide-results i(0) ${MODULE} h(pepper_desc) s("${service_string}") * i(0) i(0)


echo
echo "*** INIT INSTANCE"
set_variable TAGS C(12,buffer_size\000)
set_variable VALUES C(4,100\000)
replay 1 PPB_GetInterface:s:i  s("PPB_InputEvent;1.0") * i(1)
rpc PPP_Instance_DidCreate ${INSTANCE} i(1) ${TAGS} ${VALUES} * i(0)

replay 1 PPB_GetInterface:s:i  s("PPB_Var;1.0") * i(1)
replay 1 PPB_GetInterface:s:i  s("PPB_MouseInputEvent;1.1") * i(1)
replay 1 PPB_GetInterface:s:i  s("PPB_KeyboardInputEvent;1.0") * i(1)
# These occur once each for each event sent to the nexe
replay 7 PPB_Core_AddRefResource:i: ${EVENT} *
replay 7 ReleaseResourceMultipleTimes:ii: ${EVENT} i(1) *


echo
echo "*** SENT EVENTS EV_MOUSE_DOWN"
rpc PPP_InputEvent_HandleInputEvent ${INSTANCE} ${EVENT} ${EV_MOUSE_DOWN} ${VAR_UNDEF} * i(0)
echo
echo "*** SENT EVENTS EV_MOUSE_UP"
rpc PPP_InputEvent_HandleInputEvent ${INSTANCE} ${EVENT} ${EV_MOUSE_UP} ${VAR_UNDEF} * i(0)
echo
echo "*** SENT EVENTS EV_MOUSE_MOVE"
rpc PPP_InputEvent_HandleInputEvent ${INSTANCE} ${EVENT} ${EV_MOUSE_MOVE} ${VAR_UNDEF} * i(0)
echo
echo "*** SENT EVENTS EV_MOUSE_ENTER"
rpc PPP_InputEvent_HandleInputEvent ${INSTANCE} ${EVENT} ${EV_MOUSE_ENTER} ${VAR_UNDEF} * i(0)
echo
echo "*** SENT EVENTS EV_MOUSE_LEAVE"
rpc PPP_InputEvent_HandleInputEvent ${INSTANCE} ${EVENT} ${EV_MOUSE_LEAVE} ${VAR_UNDEF} * i(0)
echo
echo "*** SENT EVENTS EV_KEY_UP"
rpc PPP_InputEvent_HandleInputEvent ${INSTANCE} ${EVENT} ${EV_KEY_UP} ${VAR_CHAR} * i(0)
echo
echo "*** SENT EVENTS EV_KEY_DOWN"
rpc PPP_InputEvent_HandleInputEvent ${INSTANCE} ${EVENT} ${EV_KEY_DOWN} ${VAR_CHAR} * i(0)


echo
echo "*** TRIGGER EVENT FLUSH"
replay 1 PPB_GetInterface:s:i  s("PPB_Var;1.0") * i(1)
replay 1 PPB_GetInterface:s:i  s("PPB_Messaging;1.0") * i(1)
replay 1 PPB_Messaging_PostMessage:iC: i(5555) C(104,\x05\x00\x00\x00_\x00\x00\x005555:_Mouse_event:Down_modifier:left-button-down__button:Left_x:264_y:100_click_count:1_time:1\n\x00) *
replay 1 PPB_Messaging_PostMessage:iC: i(5555) C(112,\x05\x00\x00\x00b\x00\x00\x005555:_Mouse_event:Up_modifier:left-button-down__button:Left_x:264_y:100_click_count:0_time:1.0625\n\x00\x00\x00\x00\x00\x00) *
replay 1 PPB_Messaging_PostMessage:iC: i(5555) C(96,\x05\x00\x00\x00R\x00\x00\x005555:_Mouse_event:Move_modifier:_button:Left_x:264_y:100_click_count:0_time:1.125\n\x00\x00\x00\x00\x00\x00) *
replay 1 PPB_Messaging_PostMessage:iC: i(5555) C(96,\x05\x00\x00\x00T\x00\x00\x005555:_Mouse_event:Enter_modifier:_button:Left_x:264_y:100_click_count:0_time:1.1875\n\x00\x00\x00\x00) *
replay 1 PPB_Messaging_PostMessage:iC: i(5555) C(96,\x05\x00\x00\x00R\x00\x00\x005555:_Mouse_event:Leave_modifier:_button:Left_x:264_y:100_click_count:0_time:1.25\n\x00\x00\x00\x00\x00\x00) *
replay 1 PPB_Messaging_PostMessage:iC: i(5555) C(80,\x05\x00\x00\x00B\x00\x00\x005555:_Key_event:Up_modifier:_key_code:84_time:1.375_text:Var<'V'>\n\x00\x00\x00\x00\x00\x00) *
replay 1 PPB_Messaging_PostMessage:iC: i(5555) C(80,\x05\x00\x00\x00E\x00\x00\x005555:_Key_event:Down_modifier:_key_code:84_time:1.3125_text:Var<'V'>\n\x00\x00\x00) *
rpc PPP_Instance_DidChangeFocus ${INSTANCE} b(0) *


rpc PPP_ShutdownModule *

echo "DONE"
