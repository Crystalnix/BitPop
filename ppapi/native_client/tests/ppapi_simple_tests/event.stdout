
"*** REPLAY PROLOG"

"*** REGISTER REPLAY METHODS"

"*** CREATE UPCALL SERVICE"

"*** INIT MODULE"
rpc call intiated PPP_InitializeModule:iihs:ii
input 0:  i(0)
input 1:  i(4444)
input 2:  h(pepper_desc)
input 3:  @LENGTHY_SERVICE_STRING
replaying PPB_GetInterface:s:i:
input 0:  s("PPB_Core;1.0")
output 0:  i(1)
rpc call complete PPP_InitializeModule:iihs:ii


"*** INIT INSTANCE"
rpc call intiated PPP_Instance_DidCreate:iiCC:i
input 0:  i(5555)
input 1:  i(1)
input 2:  C(12,buffer_size\x00)
input 3:  C(4,100\x00)
replaying PPB_GetInterface:s:i:
input 0:  s("PPB_InputEvent;1.0")
output 0:  i(1)
rpc call complete PPP_Instance_DidCreate:iiCC:i
output 0:  i(1)


"*** SENT EVENTS EV_MOUSE_DOWN"
rpc call intiated PPP_InputEvent_HandleInputEvent:iiCC:i
input 0:  i(5555)
input 1:  i(6666)
input 2:  C(64,\x00\x00\x00\x00@\x00\x00\x00\x00\x00\x00\x00\x00\x00\xf0?\x00\x00\x00\x00\x01\x00\x00\x00\b\x01\x00\x00d\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00)
input 3:  C(8,\x00\x00\x00\x00\x00\x00\x00\x00)
replaying PPB_Core_AddRefResource:i::
input 0:  i(6666)
replaying PPB_GetInterface:s:i:
input 0:  s("PPB_MouseInputEvent;1.1")
output 0:  i(1)
replaying ReleaseResourceMultipleTimes:ii::
input 0:  i(6666)
input 1:  i(1)
replaying PPB_GetInterface:s:i:
input 0:  s("PPB_Var;1.0")
output 0:  i(1)
rpc call complete PPP_InputEvent_HandleInputEvent:iiCC:i
output 0:  i(1)


"*** SENT EVENTS EV_MOUSE_UP"
rpc call intiated PPP_InputEvent_HandleInputEvent:iiCC:i
input 0:  i(5555)
input 1:  i(6666)
input 2:  C(64,\x01\x00\x00\x00@\x00\x00\x00\x00\x00\x00\x00\x00\x00\xf1?\x00\x00\x00\x00\x00\x00\x00\x00\b\x01\x00\x00d\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00)
input 3:  C(8,\x00\x00\x00\x00\x00\x00\x00\x00)
replaying PPB_Core_AddRefResource:i::
input 0:  i(6666)
replaying ReleaseResourceMultipleTimes:ii::
input 0:  i(6666)
input 1:  i(1)
rpc call complete PPP_InputEvent_HandleInputEvent:iiCC:i
output 0:  i(1)


"*** SENT EVENTS EV_MOUSE_MOVE"
rpc call intiated PPP_InputEvent_HandleInputEvent:iiCC:i
input 0:  i(5555)
input 1:  i(6666)
input 2:  C(64,\x02\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xf2?\x00\x00\x00\x00\x00\x00\x00\x00\b\x01\x00\x00d\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00)
input 3:  C(8,\x00\x00\x00\x00\x00\x00\x00\x00)
replaying PPB_Core_AddRefResource:i::
input 0:  i(6666)
replaying ReleaseResourceMultipleTimes:ii::
input 0:  i(6666)
input 1:  i(1)
rpc call complete PPP_InputEvent_HandleInputEvent:iiCC:i
output 0:  i(1)


"*** SENT EVENTS EV_MOUSE_ENTER"
rpc call intiated PPP_InputEvent_HandleInputEvent:iiCC:i
input 0:  i(5555)
input 1:  i(6666)
input 2:  C(64,\x03\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xf3?\x00\x00\x00\x00\x00\x00\x00\x00\b\x01\x00\x00d\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00)
input 3:  C(8,\x00\x00\x00\x00\x00\x00\x00\x00)
replaying PPB_Core_AddRefResource:i::
input 0:  i(6666)
replaying ReleaseResourceMultipleTimes:ii::
input 0:  i(6666)
input 1:  i(1)
rpc call complete PPP_InputEvent_HandleInputEvent:iiCC:i
output 0:  i(1)


"*** SENT EVENTS EV_MOUSE_LEAVE"
rpc call intiated PPP_InputEvent_HandleInputEvent:iiCC:i
input 0:  i(5555)
input 1:  i(6666)
input 2:  C(64,\x04\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xf4?\x00\x00\x00\x00\x00\x00\x00\x00\b\x01\x00\x00d\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00)
input 3:  C(8,\x00\x00\x00\x00\x00\x00\x00\x00)
replaying PPB_Core_AddRefResource:i::
input 0:  i(6666)
replaying ReleaseResourceMultipleTimes:ii::
input 0:  i(6666)
input 1:  i(1)
rpc call complete PPP_InputEvent_HandleInputEvent:iiCC:i
output 0:  i(1)


"*** SENT EVENTS EV_KEY_UP"
rpc call intiated PPP_InputEvent_HandleInputEvent:iiCC:i
input 0:  i(5555)
input 1:  i(6666)
input 2:  C(64,\b\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xf6?\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00T\x00\x00\x00)
input 3:  C(16,\x05\x00\x00\x00\x01\x00\x00\x00V\x00\x00\x00\x00\x00\x00\x00)
replaying PPB_Core_AddRefResource:i::
input 0:  i(6666)
replaying PPB_GetInterface:s:i:
input 0:  s("PPB_KeyboardInputEvent;1.0")
output 0:  i(1)
replaying ReleaseResourceMultipleTimes:ii::
input 0:  i(6666)
input 1:  i(1)
rpc call complete PPP_InputEvent_HandleInputEvent:iiCC:i
output 0:  i(1)


"*** SENT EVENTS EV_KEY_DOWN"
rpc call intiated PPP_InputEvent_HandleInputEvent:iiCC:i
input 0:  i(5555)
input 1:  i(6666)
input 2:  C(64,\x07\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xf5?\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00T\x00\x00\x00)
input 3:  C(16,\x05\x00\x00\x00\x01\x00\x00\x00V\x00\x00\x00\x00\x00\x00\x00)
replaying PPB_Core_AddRefResource:i::
input 0:  i(6666)
replaying ReleaseResourceMultipleTimes:ii::
input 0:  i(6666)
input 1:  i(1)
rpc call complete PPP_InputEvent_HandleInputEvent:iiCC:i
output 0:  i(1)


"*** TRIGGER EVENT FLUSH"
rpc call intiated PPP_Instance_DidChangeFocus:ib:
input 0:  i(5555)
input 1:  b(0)
replaying PPB_GetInterface:s:i:
input 0:  s("PPB_Messaging;1.0")
output 0:  i(1)
replaying PPB_Messaging_PostMessage:iC::
input 0:  i(5555)
input 1:  C(104,\x05\x00\x00\x00_\x00\x00\x005555:_Mouse_event:Down_modifier:left-button-down__button:Left_x:264_y:100_click_count:1_time:1\n\x00)
replaying PPB_Messaging_PostMessage:iC::
input 0:  i(5555)
input 1:  C(112,\x05\x00\x00\x00b\x00\x00\x005555:_Mouse_event:Up_modifier:left-button-down__button:Left_x:264_y:100_click_count:0_time:1.0625\n\x00\x00\x00\x00\x00\x00)
replaying PPB_Messaging_PostMessage:iC::
input 0:  i(5555)
input 1:  C(96,\x05\x00\x00\x00R\x00\x00\x005555:_Mouse_event:Move_modifier:_button:Left_x:264_y:100_click_count:0_time:1.125\n\x00\x00\x00\x00\x00\x00)
replaying PPB_Messaging_PostMessage:iC::
input 0:  i(5555)
input 1:  C(96,\x05\x00\x00\x00T\x00\x00\x005555:_Mouse_event:Enter_modifier:_button:Left_x:264_y:100_click_count:0_time:1.1875\n\x00\x00\x00\x00)
replaying PPB_Messaging_PostMessage:iC::
input 0:  i(5555)
input 1:  C(96,\x05\x00\x00\x00R\x00\x00\x005555:_Mouse_event:Leave_modifier:_button:Left_x:264_y:100_click_count:0_time:1.25\n\x00\x00\x00\x00\x00\x00)
replaying PPB_Messaging_PostMessage:iC::
input 0:  i(5555)
input 1:  C(80,\x05\x00\x00\x00B\x00\x00\x005555:_Key_event:Up_modifier:_key_code:84_time:1.375_text:Var<'V'>\n\x00\x00\x00\x00\x00\x00)
replaying PPB_Messaging_PostMessage:iC::
input 0:  i(5555)
input 1:  C(80,\x05\x00\x00\x00E\x00\x00\x005555:_Key_event:Down_modifier:_key_code:84_time:1.3125_text:Var<'V'>\n\x00\x00\x00)
rpc call complete PPP_Instance_DidChangeFocus:ib:

rpc call intiated PPP_ShutdownModule::
rpc call complete PPP_ShutdownModule::

"DONE"
